File: Haskell/Algorithms/Searching-Algorithms/BinarySearch.hs
-- # Binary Search in Haskell
-- This is an implementation of simple Binary Search in Haskell on any type that can be ordered.
-- Binary Search is an effective searching algorithm, that can be used to find an element in a sorted list in O(lg(n)) time.

-- Binary search with index carry
bsWithIndex :: (Ord a) => [a] -> a -> Int -> Maybe Int
bsWithIndex list n i
    | n == head list        = Just i
    | len == 1              = Nothing   -- only candidate in list is not the right elem
    | n < head ys           = bsWithIndex xs n i
    | otherwise             = bsWithIndex ys n (i + half)
    where 
        len         = length list
        half        = len `div` 2 
        (xs, ys)    = splitAt half list

-- Binary search returning index, or -1 if element not in list
bs :: (Ord a) => [a] -> a -> Int
bs list n = case bsWithIndex list n 0 of
    Just x  -> x
    Nothing -> -1

main :: IO ()
main = do
    let intList = [1,4,7,10,25,30]
    print $ bs intList 29  -- 29 -> -1 as not in list
    print $ bs intList 7   --  7 ->  2 as in list

File: Swift/Algorithms/Searching-Algorithms/binary_search.swift
// Binary search | Swift
func binarySearch(in numbers: [Int], for value: Int) -> Int?
{
    var left = 0
    var right = numbers.count - 1

    while left <= right {

        let middle = Int(floor(Double(left + right) / 2.0))

        if numbers[middle] < value {
            left = middle + 1
        } else if numbers[middle] > value {
            right = middle - 1
        } else {
            return middle
        }
    }

    return nil
}
// Driver code
let numbers = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
let value = 17

if let index = binarySearch(in: numbers, for: value) {
    print("Found \(value) at index \(index)")
} else {
    print("Did not find \(value)")
}
